### PR TITLE
Fix moral and luck indicator messages

### PR DESCRIPTION
**Pull request by shprotru**
Fix for #2358, #2359, #2360, #2361

### CODE REVIEW

**Comment by shprotru** [comment]
Hi @LeHerosInconnu! I have some doubts on #2359 understanding, check, please this screenshot:
Снимок экрана от 2020-12-13 14-10-06
Here is hero, who robbed `Graveyard`, with army consists from undead units only.

**Comment by Branikolog** [comment]
Hi, @shprotru.
I'll take a risk answering your question. :)
I'd like to see the sentence about undead army in the bottom of this window separated by a small space from previous morale modifiers.

**Comment by shprotru** [comment]
@Branikolog new line returned back(I intended to do as in the original).
Снимок экрана от 2020-12-13 14-31-00
Снимок экрана от 2020-12-13 14-31-12

**Comment by Branikolog** [comment]
In my personal point of view, after phrase "Current Morale Modifiers" we should see all effects applied. And after that, with a small space between, we should type about the fact, that army is undead and modifiers are not applied. (as it is not morale modifier, so it is rather confusing to see it among other modifiers)
How do you think?

**Comment by shprotru** [comment]
@LeHerosInconnu, please, notice, combination for description "Current Morale Modifiers: None" was removed.

**Comment by LeHerosInconnu** [comment]
Hello @shprotru and @Branikolog,

Here's how I see it. :)

- Army without undead, without morale modifiers:
"None" must be displayed to indicate to the player that there are no active moral modifiers (it is indicated this way in the original game).

"Neutral Morale

Neutral morale means your armies will never be blessed with extra attacks or freeze in combat.

Current Morale Modifiers:

None"

- Army with living creatures and undead:

"Bad Morale

Bad morale may cause your armies to freeze in combat.

Current Morale Modifiers:

Some undead in groups -1"

When the army is composed solely of undead troops, I think it's best to first let the player know that moral modifiers are not taken into account, and only then to indicate the active morale modifiers in case living creatures would be added to the army.
(That means, the following morale modifiers are not taken into account; and then list them.)

- Army full undead, without morale modifiers:
"None" must be displayed to indicate to the player that there are no active moral modifiers.

"Neutral Morale

Neutral morale means your armies will never be blessed with extra attacks or freeze in combat.

Entire army is undead, so morale does not apply.

Current Morale Modifiers:

None"

- Army full undead, with morale modifiers:
For a full undead army, the title must always be "Neutral Morale" even if there are moral modifiers (and the neutral morale sprite displayed).

"Neutral Morale

Neutral morale means your armies will never be blessed with extra attacks or freeze in combat.

Entire army is undead, so morale does not apply.

Current Morale Modifiers:

Graveyard robber -1"

**

Perhaps the word "groups" should also be replaced by "army":

"Some undead in army -1"

Instead of:

"Some undead in groups -1"

**

In another issue, I will also make a proposal to adjust the messages for more or less positive Morale (Awful, Treason, etc.), and also for Luck. :)

- Army with living creatures and undead, and -2 to morale modifiers:

"Awful Morale

Awful morale may cause your armies to freeze in combat more often.

Current Morale Modifiers:

Some undead in army -1
Troops of 3 alignments -1"

- Army with living creatures and undead, and at least -3 to morale modifiers:

"Treason Morale

Treason morale may cause your armies to freeze in combat very often.

Current Morale Modifiers:

Some undead in army -1
Troops of 3 alignments -1
Graveyard robber -1"

**Comment by Branikolog** [comment]
Hi, @LeHerosInconnu.

> When the army is composed solely of undead troops, I think it's best to first let the player know that moral modifiers are not taken into account, and only then to indicate the active morale modifiers in case living creatures would be added to the army.

I would prefer see morale modifiers first, as experienced player would always know the origin of his army, and modifiers, as more needed information, should be higher in the window. And notification for morale not being applied, is useful for newcomers. 

> In another issue, I will also make a proposal to adjust the messages for more or less positive Morale (Awful, Treason, etc.), and also for Luck.

Totally agreed, as we have description of morale in battle by words, and it is not obvious, what this description means exactly.

**Comment by LeHerosInconnu** [comment]
Hello @Branikolog,

> Hi, @LeHerosInconnu.
> 
> > When the army is composed solely of undead troops, I think it's best to first let the player know that moral modifiers are not taken into account, and only then to indicate the active morale modifiers in case living creatures would be added to the army.
> 
> I would prefer see morale modifiers first, as experienced player would always know the origin of his army, and modifiers, as more needed information, should be higher in the window. And notification for morale not being applied, is useful for newcomers.

From the point of view of reading comprehension, it is preferable to first indicate that the morale modifiers that follow are not taken into account.

Also, putting this sentence at the end can be interpreted as being part of the morale modifiers, which is not the case (you even wrote it). :D
Therefore, it is better to put it before the "Current Morale Modifiers:" line of text. :)

**Comment by Branikolog** [comment]
> putting this sentence at the end can be interpreted as being part of the morale modifiers

That's why we make a space between modifiers and this sentence. Also, text arrangement of this window would always be the same, no matter undead army or not: there always will be description of current morale, then go modifiers, and only after optional goes this sentence. In your suggestion this sentence will appear between morale description and modifiers, which cause this window look different for this two cases.
Anyway, it's just my point of view. 
Our team can suggest their own decision. ;)

**Comment by LeHerosInconnu** [comment]
Hello @Branikolog,
 
> In your suggestion this sentence will appear between morale description and modifiers, which cause this window look different for this two cases.

This is even better. :)
"The window has changed? Yes, the army is only made up of undead, so morale modifiers are not applied."
This is a good visual mnemonic. :D

> Anyway, it's just my point of view.
> Our team can suggest their own decision. ;)

Yes. :)

File: src/fheroes2/army/army.cpp [hunk 1]
```diff
@@ -1154,13 +1154,8 @@ int Army::GetMoraleModificator( std::string * strs ) const
                 }
             }
         }
-        else {
-            if ( strs ) {
-                strs->append( _( "Entire unit is undead, so morale does not apply." ) );
-                strs->append( "\n" );
-            }
+        else
             return 0;
```
Comment (ihhub) [inline src/fheroes2/army/army.cpp:1158]: Please add `{}` to this else block.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @shprotru I left just 2 comment. Could you please take a look?

File: src/fheroes2/army/army.cpp (hunk 1 above)
Comment (shprotru) [reply src/fheroes2/army/army.cpp]: done

**Comment by ihhub** [comment]
@shprotru a bunch of thanks for fixing these issues! Great job!

**Pull request merged**